node/src/cli/validator.rs

```diff
@@ -0,0 +1,255 @@
+        &self.shared_params
+    }
+
+    fn keystore_params(&self) -> Option<&KeystoreParams> {
+        Some(&self.keystore_params)
+    }
+}
+
+/// `insert-key` subcommand.
+#[derive(Debug, Clone, Parser)]
+pub struct InsertKeyCmd {
+    #[allow(missing_docs)]
+    #[clap(flatten)]
+    pub shared_params: SharedParams,
+
+    #[allow(missing_docs)]
+    #[clap(flatten)]
+    pub keystore_params: KeystoreParams,
+
+    /// Key type
+    #[arg(long)]
+    pub key_type: String,
+
+    /// SURI
+    #[arg(long, default_value_t)]
+    pub suri: String,
+
+    /// Public key
+    #[arg(long)]
+    pub public: Bytes,
+}
+
+impl InsertKeyCmd {
+    /// Run the command
+    pub fn run<Cli: SubstrateCli>(&self, cli: &Cli) -> Result<(), Error> {
+        let keystore = init_keystore(cli, &self.shared_params, &self.keystore_params)?;
+        let key_type = self.key_type.as_str().try_into().map_err(|_| Error::KeyTypeInvalid)?;
+        keystore
+            .insert(key_type, &self.suri, &self.public[..])
+            .map_err(|_| Error::KeystoreOperation)?;
+
+        Ok(())
+    }
+}
+
+impl CliConfiguration for InsertKeyCmd {
+    fn shared_params(&self) -> &SharedParams {
+        &self.shared_params
+    }
+
+    fn keystore_params(&self) -> Option<&KeystoreParams> {
+        Some(&self.keystore_params)
+    }
+}
+
+fn init_keystore<Cli: SubstrateCli>(
+    cli: &Cli,
+    shared_params: &SharedParams,
+    keystore_params: &KeystoreParams,
+) -> Result<KeystorePtr, Error> {
+    let base_path = shared_params
+        .base_path()?
+        .unwrap_or_else(|| BasePath::from_project("", "", &Cli::executable_name()));
+    let chain_id = shared_params.chain_id(shared_params.is_dev());
+    let chain_spec = cli.load_spec(&chain_id)?;
+    let config_dir = base_path.config_dir(chain_spec.id());
+
+    match keystore_params.keystore_config(&config_dir)? {
+        KeystoreConfig::Path { path, password } => Ok(LocalKeystore::open(path, password)?.into()),
+        _ => unreachable!("keystore_config always returns path and password; qed"),
+    }
+}
+
+/// `decode-session-keys` subcommand.
+#[derive(Debug, Clone, Parser)]
+pub struct DecodeSessionKeysCmd {
+    #[allow(missing_docs)]
+    #[clap(flatten)]
+    pub shared_params: SharedParams,
+
+    #[allow(missing_docs)]
+    #[clap(flatten)]
+    pub keystore_params: KeystoreParams,
+
+    /// Hex-encoded session keys.
+    #[arg(value_name = "SESSION KEYS")]
+    pub keys: String,
+}
+
+impl DecodeSessionKeysCmd {
+    /// Run the command
+    pub fn run(&self) -> Result<(), Error> {
+        match decode_readable(&self.keys)? {
+            Some(decoded) => {
+                for key_line in decoded {
+                    println!("{}: {}", key_line.0, key_line.1);
+                }
+            },
+            None => eprintln!("Error decoding session keys"),
+        }
+
+        Ok(())
+    }
+}
+
+fn decode_readable(keys: &str) -> Result<Option<Vec<(String, String)>>, Error> {
```

Comment:
`Ok(Some(vec![]))` is valid? If not, then it's just `Result<Vec<(String, String)>>`.

Reply:
Without the context it looks like you described, but `decode_into_raw_public_keys` returns `None` on decoding failure and it does return an empty vector as well (but I don't know the real cause for this). Thus, `Option` here is for error handling actually. While there is an `Error` type for substrate CLI (which is in the result), it's external. So here we just follow the `opaque::SessionKeys::decode_into_raw_public_keys` API.

Reply:
Maybe `Option::ok_or_else` would work here to propagate this `None` to the error level?

Reply:
As I said, the `Error` is external (so we can't extend it with our case). And there is no suitable variants.

Reply:
Yeah, it has so many variants, but not _Custom_.
I did a similar thing, just pretend this error like IO.
https://github.com/Atleta-network/atleta/blob/devnet/node/src/command.rs#L256

Reply:
> Yeah, it has so many variants, but not _Custom_. I did a similar thing, just pretend this error like IO. https://github.com/Atleta-network/atleta/blob/devnet/node/src/command.rs#L256

Thanks!